File: src/mirabox/core/registry.py
"""소스와 키 등록소.

연동(integrations) 은 여기에 자기 데이터 소스와 키를 등록한다. 코어는
무엇이 등록됐는지만 알 뿐 각 연동의 사정은 모른다. 연동을 하나 추가하는
일이 파일 하나 추가로 끝나게 하려는 구조다.

키는 자기가 필요한 소스를 선언한다. 데몬은 보드에 올라온 키가 요구하는
소스만 켠다. MAIL 키를 안 쓰면 Gmail 을 아예 호출하지 않는다.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from PIL import Image

from .state import State

FetchFn = Callable[[], Any]
RenderFn = Callable[[int, State, dict], Image.Image]


@dataclass(frozen=True)
class Source:
    name: str
    fetch: FetchFn
    every: float          # 초


@dataclass(frozen=True)
class Option:
    """키마다 다른 개별 설정. 조작 화면이 이걸 보고 입력란을 만든다."""
    name: str
    label: str
    kind: str = "text"        # text 또는 file
    placeholder: str = ""


@dataclass(frozen=True)
class Key:
    name: str
    label: str            # 키에 찍히는 짧은 이름
    summary: str          # 조작 화면에 보여줄 한 줄 설명
    render: RenderFn
    sources: frozenset[str]
    options: tuple[Option, ...] = ()


SOURCES: dict[str, Source] = {}
KEYS: dict[str, Key] = {}


def source(name: str, every: float) -> Callable[[FetchFn], FetchFn]:
    """데이터 소스를 등록한다. fetch 는 실패하면 예외를 던지면 된다."""
    def register(fn: FetchFn) -> FetchFn:
        SOURCES[name] = Source(name, fn, every)
        return fn
    return register


def key(name: str, label: str, summary: str,
        sources: tuple[str, ...] = (),
        options: tuple[Option, ...] = ()) -> Callable[[RenderFn], RenderFn]:
    def register(fn: RenderFn) -> RenderFn:
        KEYS[name] = Key(name, label, summary, fn, frozenset(sources), options)
        return fn
    return register


def sources_for(key_names) -> set[str]:
    """이 키들을 그리는 데 필요한 소스 이름."""
    needed: set[str] = set()
    for name in key_names:
        entry = KEYS.get(name)
        if entry:
            needed |= entry.sources
    return needed

File: src/main/registry.ts
/**
 * 소스와 키 등록소.
 *
 * 연동(integrations)은 여기에 자기 데이터 소스와 키를 등록한다. 코어는
 * 무엇이 등록됐는지만 알 뿐 각 연동의 사정은 모른다. 연동을 하나 추가하는
 * 일이 파일 하나 추가로 끝나게 하려는 구조다.
 *
 * 키는 자기가 필요한 소스를 선언한다. 데몬은 보드에 올라온 키가 요구하는
 * 소스만 켠다. 메일 키를 안 쓰면 Gmail 을 아예 호출하지 않는다.
 */

import type { Canvas } from '@napi-rs/canvas'
import type { Family } from './render.js'
import type { KeyOption } from '../shared/types.js'

export interface State {
  data: Record<string, unknown>
  errors: Record<string, string>
  updatedAt: Record<string, number>
}

export const emptyState = (): State => ({ data: {}, errors: {}, updatedAt: {} })

export function pick<T>(state: State, source: string): T | null {
  return (state.data[source] as T | undefined) ?? null
}

export interface Source {
  name: string
  /** 실패하면 던지면 된다. 수집기가 잡아서 오류로 남긴다. */
  fetch: () => Promise<unknown>
  /** 초 */
  every: number
  /**
   * 값 자체가 낡았을 때 조작 화면에 띄울 한 줄.
   *
   * 마지막으로 가져온 시각과 값이 만들어진 시각은 다르다. 계정 한도처럼 남이
   * 떨궈 준 파일을 읽는 소스는 방금 읽었어도 내용이 두 시간 전 것일 수 있다.
   */
  describe?: (value: unknown) => string | null
}

export type RenderFn = (
  index: number,
  state: State,
  options: Record<string, string>,
) => Canvas

export interface Key {
  name: string
  /** 키에 찍히는 짧은 이름 */
  label: string
  /** 조작 화면에 보여줄 한 줄 설명 */
  summary: string
  /** 어디서 온 키인지. 조작 화면의 묶음과 기기 표식이 여기서 나온다. */
  family: Family
  render: RenderFn
  sources: string[]
  options: KeyOption[]
}

export const SOURCES = new Map<string, Source>()
export const KEYS = new Map<string, Key>()

export function source(
  name: string,
  every: number,
  fetch: () => Promise<unknown>,
  describe?: (value: unknown) => string | null,
): void {
  SOURCES.set(name, { name, every, fetch, describe })
}

export function key(spec: {
  name: string
  label: string
  summary: string
  family?: Family
  sources?: string[]
  options?: KeyOption[]
  render: RenderFn
}): void {
  KEYS.set(spec.name, {
    name: spec.name,
    label: spec.label,
    summary: spec.summary,
    family: spec.family ?? 'system',
    render: spec.render,
    sources: spec.sources ?? [],
    options: spec.options ?? [],
  })
}

/** 이 키들을 그리는 데 필요한 소스 이름. */
export function sourcesFor(keyNames: Iterable<string>): Set<string> {
  const needed = new Set<string>()
  for (const name of keyNames) {
    for (const source of KEYS.get(name)?.sources ?? []) needed.add(source)
  }
  return needed
}
